### PR TITLE
[US041] Add Authorisation System To API

### PR DESCRIPTION
**Pull request by TaylorJCrawford**

### CODE REVIEW

**Review by DarraghHaughey-Kainos — APPROVED** [review]
great :)

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -41,9 +41,11 @@ jobs:
           DB_PASSWORD: ${{ secrets.DB_PASSWORD }}
           DB_NAME: ${{ secrets.DB_NAME }}
           JWT_SECRET: ${{ secrets.JWT_SECRET }}
+          TEST_VALID_EMAIL_API: ${{ secrets.TEST_VALID_EMAIL_API }}
```
Comment (glenn-hamilton) [inline .github/workflows/docker.yml:44]: These need to go in the `Dockerfile` as well

File: .github/workflows/maven.yml [hunk 2]
```diff
@@ -32,4 +32,6 @@ jobs:
           DB_USERNAME: ${{ secrets.DB_USERNAME }}
           DB_PASSWORD: ${{ secrets.DB_PASSWORD }}
           JWT_SECRET: ${{ secrets.JWT_SECRET }}
-        run: mvn -DJWT_SECRET=${JWT_SECRET} -DDB_HOST=${DB_HOST} -DDB_NAME=${DB_NAME} -DDB_USERNAME=${DB_USERNAME} -DDB_PASSWORD=${DB_PASSWORD} -B package --file pom.xml
+          TEST_VALID_EMAIL_API: ${{ secrets.TEST_VALID_EMAIL_API }}
```
Comment (glenn-hamilton) [inline .github/workflows/maven.yml:35]: These need to go in the `Dockerfile` as well

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -32,4 +32,6 @@ jobs:
           DB_USERNAME: ${{ secrets.DB_USERNAME }}
           DB_PASSWORD: ${{ secrets.DB_PASSWORD }}
           JWT_SECRET: ${{ secrets.JWT_SECRET }}
-        run: mvn -DJWT_SECRET=${JWT_SECRET} -DDB_HOST=${DB_HOST} -DDB_NAME=${DB_NAME} -DDB_USERNAME=${DB_USERNAME} -DDB_PASSWORD=${DB_PASSWORD} clean test
\ No newline at end of file
+          TEST_VALID_EMAIL_API: ${{ secrets.TEST_VALID_EMAIL_API }}
```
Comment (glenn-hamilton) [inline .github/workflows/test.yml:35]: These need to go in the `Dockerfile` as well

File: src/main/java/org/kainos/ea/resources/CapabilityController.java [hunk 4]
```diff
@@ -16,23 +19,31 @@
 public class CapabilityController {
 
     private final CapabilityService capabilityService;
+    private AuthService authService;
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/resources/CapabilityController.java]: This can be `final`

Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/resources/CapabilityController.java]: Done :)

File: .github/workflows/test.yml (hunk 3 above)
Comment (TaylorJCrawford) [reply .github/workflows/test.yml:35]: Done :)

File: .github/workflows/maven.yml (hunk 2 above)
Comment (TaylorJCrawford) [reply .github/workflows/maven.yml:35]: Done :)

File: .github/workflows/docker.yml (hunk 1 above)
Comment (TaylorJCrawford) [reply .github/workflows/docker.yml:44]: Done :)

File: src/main/java/org/kainos/ea/api/AuthService.java [hunk 5]
```diff
@@ -23,22 +28,45 @@ public AuthService(DatabaseConnector databaseConnector, AuthDao authDao, Credent
...
+    public Claims isValidToken(String token, String requiredPermission) throws AuthenticationException {
         if (token == null) {
             throw new AuthenticationException("No token provided");
+        } else if (requiredPermission == null) {
+            throw new AuthenticationException("No token permission set for endpoint");
         }
-        return authDao.parseToken(token);
+
+        Claims decodedJWT = authDao.parseToken(token);
+        System.out.println(decodedJWT);
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/api/AuthService.java]: Don't need to log the decoded JWT here

**Review by glenn-hamilton — APPROVED** [review]

File: src/main/java/org/kainos/ea/api/AuthService.java (hunk 5 above)
Comment (TaylorJCrawford) [reply src/main/java/org/kainos/ea/api/AuthService.java]: Fixed :)

File: src/test/java/com/kainos/ea/integration/CapabilityIntegrationTest.java [hunk 6]
```diff
@@ -20,13 +28,31 @@ public class CapabilityIntegrationTest {
             new ResourceConfigurationSourceProvider()
     );
 
+    String validToken;
+
+    @BeforeEach
+    public void init() {
+        String fakeSecret = System.getenv("JWT_SECRET");
```
Comment (glenn-hamilton) [inline src/test/java/com/kainos/ea/integration/CapabilityIntegrationTest.java]: Is this really fake if it's using the `JWT_SECRET` environment variable 🤔

Comment (TaylorJCrawford) [reply src/test/java/com/kainos/ea/integration/CapabilityIntegrationTest.java]: All fixed :)

File: src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 7]
```diff
@@ -26,11 +37,50 @@ public class JobRoleIntegrationTest {
     );
 
     @Test
-    void getJobRoles_shouldReturnListOfJobRoles() {
-        List<JobRole> response = APP.client().target("http://localhost:8080/api/job-roles")
+    void getJobRoles_shouldReturn200_whenValidRoleIsPassedIn () {
+
+        String fakeSecret = System.getenv("JWT_SECRET");
```
Comment (glenn-hamilton) [inline src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: fakeSecret that isn't fake

File: src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 8]
```diff
@@ -26,11 +37,50 @@ public class JobRoleIntegrationTest {
...
+                .header("Authorization", token)
+                .get();
+
+        Assertions.assertEquals(200, response.getStatus());
+    }
+
+    @Test
+    void getJobRoles_shouldReturn401_whenInvalidRoleIsPassedIn () {
+
+        String fakeSecret = System.getenv("JWT_SECRET");
```
Comment (glenn-hamilton) [inline src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: fakeSecret that isn't fake

Comment (TaylorJCrawford) [reply src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: shshs - > They will never know. All fixed :)

File: src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java (hunk 7 above)
Comment (TaylorJCrawford) [reply src/test/java/com/kainos/ea/integration/JobRoleIntegrationTest.java]: Fixed :)

**Pull request merged**